File: pipeline/extractors/text_reader.py
import os
import re

def read_plain_text_file(filepath, default_genre=None):
    """
    Reads plain text or markdown files (.txt, .md).
    Detects genre if default_genre is auto or None (prose vs poetry vs drama).
    Returns list of element dicts:
      {'type': 'heading'|'paragraph'|'verse_line'|'stanza_break'|'stage_direction'|'character_name',
       'body': text, 'speaker': name_if_any}
    """
    with open(filepath, 'r', encoding='utf-8', errors='ignore') as fp:
        lines = fp.readlines()
        
    elements = []
    
    # Check genre auto-detection hints
    has_verse_stanzas = False
    blank_lines = 0
    short_lines = 0
    drama_character_lines = 0
    
    for line in lines:
        stripped = line.strip()
        if not stripped:
            blank_lines += 1
        else:
            if len(stripped) < 50:
                short_lines += 1
            if re.match(r'^[A-ZА-ЯЁ\s]{2,30}[:.]', stripped):
                drama_character_lines += 1

    detected_genre = default_genre
    if not detected_genre or detected_genre == 'auto':
        if drama_character_lines > 5 and drama_character_lines / max(1, len(lines)) > 0.1:
            detected_genre = 'drama'
        elif short_lines > 10 and blank_lines > 5 and (short_lines + blank_lines) / max(1, len(lines)) > 0.5:
            detected_genre = 'poetry'
        else:
            detected_genre = 'prose'
            
    # Parse elements based on detected_genre
    in_stanza = False
    
    for idx, line in enumerate(lines):
        raw_line = line.rstrip('\r\n')
        stripped = raw_line.strip()
        
        if not stripped:
            if detected_genre == 'poetry':
                elements.append({'type': 'stanza_break', 'body': ''})
            continue
            
        # Check headings / chapter markers
        if re.match(r'^(#+\s*|Глава\s+\d+|Часть\s+\d+|Акт\s+\d+|Сцена\s+\d+|Chapter\s+\d+)', stripped, re.IGNORECASE):
            clean_heading = re.sub(r'^#+\s*', '', stripped)
            elements.append({'type': 'heading', 'body': clean_heading})
            continue
            
        if detected_genre == 'poetry':
            elements.append({'type': 'verse_line', 'body': stripped})
        elif detected_genre == 'drama':
            # Check character name line e.g. "КИР:" or "АНФИЯ (входит):"
            char_match = re.match(r'^([A-ZА-ЯЁ\s]{2,30})\s*(?:\((.*?)\))?[:.](.*)$', stripped)
            if char_match:
                cname, sdir, dialogue = char_match.groups()
                elements.append({'type': 'character_name', 'body': cname.strip(), 'speaker': cname.strip()})
                if sdir:
                    elements.append({'type': 'stage_direction', 'body': f"({sdir.strip()})"})
                if dialogue.strip():
                    elements.append({'type': 'paragraph', 'body': dialogue.strip(), 'speaker': cname.strip()})
            elif stripped.startswith('(') and stripped.endswith(')'):
                elements.append({'type': 'stage_direction', 'body': stripped})
            else:
                elements.append({'type': 'paragraph', 'body': stripped})
        else:  # prose
            elements.append({'type': 'paragraph', 'body': stripped})
            
    return elements, detected_genre

File: src/extractors/text_reader.rs
use std::fs;
use std::path::Path;
use regex::Regex;
use encoding_rs::WINDOWS_1251;
use crate::models::{ElementType, Genre, LiteratureElement};

pub fn read_plain_text<P: AsRef<Path>>(filepath: P, default_genre: Genre) -> (Vec<LiteratureElement>, Genre) {
    let bytes = match fs::read(&filepath) {
        Ok(b) => b,
        Err(_) => return (Vec::new(), default_genre),
    };

    if bytes.is_empty() {
        return (Vec::new(), default_genre);
    }

    // 1. Strip UTF-8 BOM if present
    let raw_bytes = if bytes.starts_with(&[0xEF, 0xBB, 0xBF]) {
        &bytes[3..]
    } else {
        &bytes[..]
    };

    // 2. Multi-encoding decoding (UTF-8 -> Windows-1251 -> Lossy)
    let content = match std::str::from_utf8(raw_bytes) {
        Ok(s) => s.to_string(),
        Err(_) => {
            let (decoded, _, has_errors) = WINDOWS_1251.decode(raw_bytes);
            if !has_errors {
                decoded.into_owned()
            } else {
                String::from_utf8_lossy(raw_bytes).into_owned()
            }
        }
    };

    let lines: Vec<String> = content.lines().map(|l| l.to_string()).collect();

    let mut blank_lines = 0;
    let mut short_lines = 0;
    let mut drama_lines = 0;

    let re_heading = Regex::new(r"(?i)^(?:#+\s*|Глава\s+\d+|Часть\s+\d+|Акт\s+\d+|Сцена\s+\d+|Chapter\s+\d+)").unwrap();
    let re_drama_char = Regex::new(r"^([A-ZА-ЯЁ\s]{2,30})\s*(?:\((.*?)\))?[:.](.*)$").unwrap();

    for line in &lines {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            blank_lines += 1;
        } else {
            if trimmed.len() < 50 {
                short_lines += 1;
            }
            if re_drama_char.is_match(trimmed) {
                drama_lines += 1;
            }
        }
    }

    let detected_genre = match default_genre {
        Genre::Auto => {
            let total = lines.len().max(1);
            if drama_lines > 3 && (drama_lines as f64 / total as f64) > 0.05 {
                Genre::Drama
            } else if short_lines > 5 && (short_lines + blank_lines) as f64 / total as f64 > 0.4 {
                Genre::Poetry
            } else {
                Genre::Prose
            }
        }
        g => g,
    };

    let mut elements = Vec::new();

    for line in lines {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            if detected_genre == Genre::Poetry {
                elements.push(LiteratureElement {
                    element_type: ElementType::StanzaBreak,
                    body: String::new(),
                    edited_body: String::new(),
                    speaker: None,
                });
            }
            continue;
        }

        if re_heading.is_match(trimmed) {
            elements.push(LiteratureElement {
                element_type: ElementType::Heading,
                body: trimmed.to_string(),
                edited_body: trimmed.to_string(),
                speaker: None,
            });
            continue;
        }

        match detected_genre {
            Genre::Poetry => {
                elements.push(LiteratureElement {
                    element_type: ElementType::VerseLine,
                    body: trimmed.to_string(),
                    edited_body: trimmed.to_string(),
                    speaker: None,
                });
            }
            Genre::Drama => {
                if let Some(caps) = re_drama_char.captures(trimmed) {
                    let cname = caps.get(1).unwrap().as_str().trim().to_string();
                    let sdir = caps.get(2).map(|m| m.as_str().trim().to_string());
                    let dialogue = caps.get(3).unwrap().as_str().trim().to_string();

                    elements.push(LiteratureElement {
                        element_type: ElementType::CharacterName,
                        body: cname.clone(),
                        edited_body: cname.clone(),
                        speaker: Some(cname.clone()),
                    });

                    if let Some(sd) = sdir {
                        elements.push(LiteratureElement {
                            element_type: ElementType::StageDirection,
                            body: format!("({})", sd),
                            edited_body: format!("({})", sd),
                            speaker: Some(cname.clone()),
                        });
                    }

                    if !dialogue.is_empty() {
                        elements.push(LiteratureElement {
                            element_type: ElementType::Paragraph,
                            body: dialogue.clone(),
                            edited_body: dialogue,
                            speaker: Some(cname),
                        });
                    }
                } else if trimmed.starts_with('(') && trimmed.ends_with(')') {
                    elements.push(LiteratureElement {
                        element_type: ElementType::StageDirection,
                        body: trimmed.to_string(),
                        edited_body: trimmed.to_string(),
                        speaker: None,
                    });
                } else {
                    elements.push(LiteratureElement {
                        element_type: ElementType::Paragraph,
                        body: trimmed.to_string(),
                        edited_body: trimmed.to_string(),
                        speaker: None,
                    });
                }
            }
            _ => {
                elements.push(LiteratureElement {
                    element_type: ElementType::Paragraph,
                    body: trimmed.to_string(),
                    edited_body: trimmed.to_string(),
                    speaker: None,
                });
            }
        }
    }

    (elements, detected_genre)
}
